roachtest: jasync failed

roachtest.jasync [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/8641928?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/8641928?buildTab=artifacts#/jasync) on master @ [09188370d82e163ff1d44c62fe611104502c548d](https://github.com/cockroachdb/cockroach/commits/09188370d82e163ff1d44c62fe611104502c548d):


```
test artifacts and logs in: /artifacts/jasync/run_1
(orm_helpers.go:191).summarizeFailed: 
Tests run on Cockroach v23.1.0-alpha.1-1805-g09188370d8
Tests run against jasyncsql 6301aa1b9ef8a0d4c5cf6f3c095b30a388c62dc0
158 Total Tests Run
82 tests passed
76 tests failed
5 tests skipped
0 tests ignored
0 tests passed unexpectedly
21 tests failed unexpectedly
0 tests expected failed but skipped
0 tests expected failed but not run
---
--- FAIL: com.github.aysnc.sql.db.integration.pool.SingleThreadedAsyncObjectPoolSpec.pool should exhaust the pool - unknown (unexpected)
--- FAIL: com.github.aysnc.sql.db.integration.TimeAndDateSpec.when processing times and dates should handle sending a LocalDateTime and return a LocalDateTime for a timestamp without timezone column - unknown (unexpected)
--- FAIL: com.github.aysnc.sql.db.integration.TimeAndDateSpec.when processing times and dates should handle sending a date with timezone and retrieving the date , the same time zone - unknown (unexpected)
--- FAIL: com.github.aysnc.sql.db.integration.BitSpec.when processing bit columns should result in binary data in BIT(2) column - unknown (unexpected)
--- FAIL: com.github.aysnc.sql.db.integration.pool.SuspendingPoolSpec.transactions should rollback on error - unknown (unexpected)
--- FAIL: com.github.aysnc.sql.db.integration.TimeAndDateSpec.when processing times and dates should support a time object - unknown (unexpected)
--- FAIL: com.github.aysnc.sql.db.integration.pool.ConnectionPoolSpec.pool should runs commands for a transaction in a single connection - unknown (unexpected)
--- FAIL: com.github.aysnc.sql.db.integration.TimeAndDateSpec.when processing times and dates should support timestamp with timezone and microseconds - unknown (unexpected)
--- FAIL: com.github.aysnc.sql.db.integration.pool.ActorAsyncObjectPoolSpec.pool should it should validate returned connections before sending them back to the pool - unknown (unexpected)
--- FAIL: com.github.aysnc.sql.db.integration.TimeAndDateSpec.when processing times and dates should support intervals - unknown (unexpected)
--- FAIL: com.github.aysnc.sql.db.integration.pool.NextGenConnectionPoolSpec.pool should runs commands for a transaction in a single connection - unknown (unexpected)
--- FAIL: com.github.aysnc.sql.db.integration.BitSpec.when processing bit columns should result in binary data - unknown (unexpected)
--- FAIL: com.github.aysnc.sql.db.integration.TimeAndDateSpec.when processing times and dates should supports sending a local date and later a date time object for the same field - unknown (unexpected)
--- FAIL: com.github.aysnc.sql.db.integration.TimeAndDateSpec.when processing times and dates should handle sending a time with timezone and return a LocalDateTime for a timestamp without timezone column - unknown (unexpected)
--- FAIL: com.github.aysnc.sql.db.integration.TimeAndDateSpec.when processing times and dates should support timestamp with timezone - unknown (unexpected)
--- FAIL: com.github.aysnc.sql.db.integration.pool.SingleThreadedAsyncObjectPoolSpec.pool should it should validate returned connections before sending them back to the pool - unknown (unexpected)
--- FAIL: com.github.aysnc.sql.db.integration.TimeAndDateSpec.when processing times and dates should support current_timestamp with timezone - unknown (unexpected)
--- FAIL: com.github.aysnc.sql.db.integration.pool.SingleThreadedAsyncObjectPoolSpec.pool should it should not accept returned connections that aren't ready for query - unknown (unexpected)
--- FAIL: com.github.aysnc.sql.db.integration.pool.ActorAsyncObjectPoolSpec.pool should it should not accept returned connections that aren't ready for query - unknown (unexpected)
--- FAIL: com.github.aysnc.sql.db.integration.TimeAndDateSpec.when processing times and dates should support a time object, microseconds - unknown (unexpected)
--- FAIL: com.github.aysnc.sql.db.integration.pool.SuspendingPoolSpec.transactions with pool should rollback on error - unknown (unexpected)
For a full summary look at the jasyncsql artifacts
An updated blocklist (jasyncsqlBlocklist23_1) is available in the artifacts' jasyncsql log
```
<p>Parameters: <code>ROACHTEST_cloud=gce</code>
, <code>ROACHTEST_cpu=4</code>
, <code>ROACHTEST_encrypted=false</code>
, <code>ROACHTEST_ssd=0</code>
</p>
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #95816 roachtest: jasync failed [C-test-failure O-roachtest O-robot T-sql-sessions branch-release-22.1]
</p>
</details>
/cc @cockroachdb/sql-sessions
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*jasync.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-24369

--- a/pkg/sql/logictest/testdata/logic_test/timestamp
+++ b/pkg/sql/logictest/testdata/logic_test/timestamp
@@ -656,6 +656,24 @@ SELECT to_char(d, 'FF1 FF2 FF3 FF4 FF5 FF6  ff1 ff2 ff3 ff4 ff5 ff6  MS US')
 7 78 789 7890 78901 789010  7 78 789 7890 78901 789010  789 789010
 7 78 789 7890 78901 789012  7 78 789 7890 78901 789012  789 789012
 
+query T
+SET timezone = '00:00';
+SELECT ARRAY[
+  '2018-11-02 12:34:56.78901234'::timestamptz,
+  '2018-11-02 12:34:56.78901234-07'::timestamptz
+]
+----
+{"2018-11-02 12:34:56.789012+00","2018-11-02 19:34:56.789012+00"}
+
+query T
+SET timezone = 'Australia/Sydney';
+SELECT ARRAY[
+  '2018-11-02 12:34:56.78901234'::timestamptz,
+  '2018-11-02 12:34:56.78901234-07'::timestamptz
+]
+----
+{"2018-11-02 12:34:56.789012+11","2018-11-03 06:34:56.789012+11"}
+
 query TT
 SET timezone = '00:00';
 SELECT to_char(now(), 'OF') as of_t, to_char(now(), 'TZH:TZM') as "TZH:TZM";

--- a/pkg/sql/pgwire/conn.go
+++ b/pkg/sql/pgwire/conn.go
@@ -1442,7 +1442,7 @@ func (c *tenantEgressCounter) GetBatchNetworkEgress(
 			// Use the default values for the DataConversionConfig and location.
 			// See the comment in getRowNetworkEgress for why the writeText variant
 			// is used here instead of writeBinary.
-			c.buf.writeTextColumnarElement(ctx, &c.vecs, vecIdx, rowIdx, conv, nil /* sessionLoc */)
+			c.buf.writeTextColumnarElement(ctx, &c.vecs, vecIdx, rowIdx, conv, time.UTC)
 			egress += int64(c.buf.Len())
 			c.buf.reset()
 		}

--- a/pkg/sql/pgwire/types.go
+++ b/pkg/sql/pgwire/types.go
@@ -157,8 +157,13 @@ func writeTextDatumNotNull(
 	sessionLoc *time.Location,
 	t *types.T,
 ) {
+
 	oldDCC := b.textFormatter.SetDataConversionConfig(conv)
-	defer b.textFormatter.SetDataConversionConfig(oldDCC)
+	oldLoc := b.textFormatter.SetLocation(sessionLoc)
+	defer func() {
+		b.textFormatter.SetDataConversionConfig(oldDCC)
+		b.textFormatter.SetLocation(oldLoc)
+	}()
 	switch v := tree.UnwrapDOidWrapper(d).(type) {
 	case *tree.DBitArray:
 		b.textFormatter.FormatNode(v)
@@ -293,7 +298,11 @@ func (b *writeBuffer) writeTextColumnarElement(
 	sessionLoc *time.Location,
 ) {
 	oldDCC := b.textFormatter.SetDataConversionConfig(conv)
-	defer b.textFormatter.SetDataConversionConfig(oldDCC)
+	oldLoc := b.textFormatter.SetLocation(sessionLoc)
+	defer func() {
+		b.textFormatter.SetDataConversionConfig(oldDCC)
+		b.textFormatter.SetLocation(oldLoc)
+	}()
 	typ := vecs.Vecs[vecIdx].Type()
 	if log.V(2) {
 		log.Infof(ctx, "pgwire writing TEXT columnar element of type: %s", typ)

--- a/pkg/sql/sem/tree/format.go
+++ b/pkg/sql/sem/tree/format.go
@@ -361,6 +361,13 @@ func (ctx *FmtCtx) SetDataConversionConfig(
 	return old
 }
 
+// SetLocation sets the location on ctx and returns the old one.
+func (ctx *FmtCtx) SetLocation(loc *time.Location) *time.Location {
+	old := ctx.location
+	ctx.location = loc
+	return old
+}
+
 // WithReformatTableNames modifies FmtCtx to to substitute the printing of table
 // names using the provided function, calls fn, then restores the original table
 // formatting.

